### DISCUSSION TITLE
updating the integration loses sensor Cost data

### DISCUSSION THREAD

**Discussion by insipiens** [wills106/homeassistant-solax-modbus#1011] [General]
Hi

When upgrading to the latest version when there are breaking changes it is suggested to delete the current configuration and then re-enter the configuration to retain the historic data relating to the entities.

This seems to work fine but in one case it doesn't - that's when you have a sensor Cost recording the usage, in my case I have the octopus integration providing unit costs in tandem with the inverter data - so Today's export creates sensor Compensation, and today's import creates the sensor Cost value. 

Whichever way I try and update solax modbus  I always lose the historic costs which in my case is more than a year's worth of daily 
data.

Has anyone also found this, or found a way of upgrading without the loss?

Thanks

John

**Comment by pesi78** [comment]
I have the same issue. After updating from 2024.05.7 to 2024.08.12 the history costs are away from the energy dashboard.

**Comment by felge20000** [comment]
Same issue here after the update (with removing the inverter prior and re-adding it after).
All other data was kept except the calculation of the cost / gains.
I'm using the default energy panel of home assistant. The kWh values are correct (all of 2024), the € values sadly only the ones since the update a few days ago.
The data is still there to calculate, so hoping for a fix :)

Edit: This is a discussion thread right now. Imho this is a proper _issue_

![solax-screen-no-cost-data](https://github.com/user-attachments/assets/cf33c74b-8c3a-428c-ae20-e754dba304f7)

**Reply by insipiens** [reply]
The reason I posted here as a discussion is that it most likely isn’t caused by this integration but by HA core. In that case we would have to raise it as an issue with HA. 

The process outlined for upgrading the integration is correct, you can find it on the HA discord energy channel as a faq.

I was probably lucky in having daily back ups, so just restored and haven’t upgraded since….

My motivation was to figure out a work around at least….but strange, could it be only be 3 of us have experienced this? Maybe you’re right. 

@wills106 Do you have some technical insight into this? At least something that will lead us towards understanding the problem? My gut feeling is its something to do with the underlaying entity Ids, a numeric Id….

cheers

John

**Comment by insipiens** [comment]
I put this script together to check the cost/comp statistics that are stored in the sql lite DB

```
SELECT statistics_meta.statistic_id,statistics.id,count(*),strftime('%d - %m - %Y ', datetime(min(created_ts), 'unixepoch')),count(*),unit_of_measurement
FROM "statistics_meta" 
left outer join statistics on statistics.metadata_id = statistics_meta.id
where 
statistics_meta.statistic_id like "sensor.SolaX%"
AND unit_of_measurement = "GBP"
group by statistics_meta.id
```

I found that it had renamed the sensors and the old values still existed, I had added the integration on the 23rd August last year.

![image](https://github.com/user-attachments/assets/89f8b1ea-2b50-4ec1-b4f0-06b06b6f5fd0)


Wondering if its possible to merge the stats together

**Reply by insipiens** [reply]
I think I sort of solved this by editing the entity registry [/homaeassistant/.storage/core.entity_registry] and changing the 4 additional cost compensation entries,solax_today_s_export_energy_cost_3 etc. either removing their _2 and _3 in the entity names. So changing it to this on both entries (cost and compensation),sorry didn't screenshot the before bit: 




Then I deleted the solax modbus config and restarted HA. Not sure deleting the integration config is required......

On restart at this point you should get the history showing again from the stat/DB entries without the underscore.

Reconfigure the integration and check if it starts recording the stats again. it takes a while as the statistics update in 5 minute blocks

If you try this please make a good restorable back up first - I'm taking no responsibility for this failing! This is clearly a hack and there may be downsides that I haven't seen yet on my system like system integrity.

What I think is happening is that removing the _2 or _3 in the  additional sensors in the registry we're repointing the in use stats back to the oldest one, there's only one of each entry in the registry for the solax import/export sensors. So maybe the restart is all that is required...haven't tested it yet  - going for a lie down instead ;-) 

If you do try it let us know how it goes

Insipiens

PS saying solved is probably doing a lot of lifting here
